### ISSUE TITLE
Benchmark available for PTV3?

### ISSUE THREAD

**Issue by mikel-zhobro** [NVlabs/WarpConvNet#15]
Thank you the nice library!

I am trying to figure out the advantages of WarpConvNet's PTV3 implementation w.r.t the default PTV3 implementation.

1. Is there any benchmarking (memory & time) script for PTV3 or any of its components? 
2. when is `wp.init()` used? For example, in [test_attention.py](https://github.com/NVlabs/WarpConvNet/blob/2cde94fadd5cab9b9685aa35b52dd340f05e61f3/tests/nn/test_attention.py#L31-L41) it is used whereas in [point_transformerv3.py](https://github.com/NVlabs/WarpConvNet/blob/2cde94fadd5cab9b9685aa35b52dd340f05e61f3/examples/point_transformer_v3.py#L276-L285) it is not.

**Comment by kasparas-k** [comment]
1. I did some memory benchmarking of PTV3 in the official [Pointcept](https://github.com/Pointcept/Pointcept) and the one in [examples](https://github.com/NVlabs/WarpConvNet/blob/main/examples/point_transformer_v3.py)

I used Pytorch Lightning to do a basic training loop with bf16 precision to match the out of the box capabilities of Pointcept as close as possible. This test was done with a real point cloud building batches consisting of copies of the exact same data sample to reduce memory up-and-down due to sparse tensors having different sizes.

**WarpConvNet example vs Pointcept**

Despite the default configs being similar, the WarpConvNet version has some significant differences in its implementation. leading to an additional 4M parameters and almost double VRAM for the same data, compared to Pointcept. But it's not a fair comparison since it's a bigger model, of course it will take more VRAM.

**Architecture-matched WarpConvNet vs Pointcept**

First change is in the PatchAttentionBlock, if you look at the original implementation you'll see that the self.mlp layer is just that, two MLP in succession
https://github.com/Pointcept/Pointcept/blob/e298a7d9ad7b06c5d4f5de15654d6d4df70af36e/pointcept/models/point_transformer_v3/point_transformer_v3m1_base.py#L225-L248

The FeedForward from WarpConvNet has a gating mechanism with a total of three linear layers.
https://github.com/NVlabs/WarpConvNet/blob/12e83127f2fcbaefb67d2c6d892f2dd49f7ab727/warpconvnet/nn/modules/attention.py#L450-L468

Second change is the unpooling block: the official implementation projects the feature and skip feature separately then adds them:
https://github.com/Pointcept/Pointcept/blob/e298a7d9ad7b06c5d4f5de15654d6d4df70af36e/pointcept/models/point_transformer_v3/point_transformer_v3m1_base.py#L471-L482

while in the WarpConvNet example feature is projected, then concatenated wtih skip feature and they are projected together again:
https://github.com/NVlabs/WarpConvNet/blob/12e83127f2fcbaefb67d2c6d892f2dd49f7ab727/examples/point_transformer_v3.py#L186-L199

Finally a petty change for the last 32 parameters, the initial embedding SparseConvolution in Pointcept has `bias=False`.

After switching out these layers I get the parameter counts in both models to be exactly 46'154'272 (in contrast to the original example's 50'786'176).

How does VRAM usage compare when the parameter counts are matched? I used the method from this [blog post](https://huggingface.co/blog/train_memory)

Pointcept:

<img width="1421" height="596" alt="Image" src="https://github.com/user-attachments/assets/08675b6b-87e2-4612-9f38-abf71c67e8bc" />

WarpConvNet + Lightning:

<img width="1420" height="586" alt="Image" src="https://github.com/user-attachments/assets/bc143f5c-14d3-4eed-8c19-21ff2e136aa8" />

In Pointcept, peak VRAM usage is roughly 16 GB for a batch size of four 8 m cylindrical point clouds with ~200k voxels filled, while for the same data and same parameter count WarpConvNet + Lightning takes 22 GB, a very significant increase in VRAM usage. Also, because my GPU profile record is limited lenght, you can see that the full history is longer for Pointcept, so that means it's not that there are larger allocations of memory, but more allocations. I'm not sure why that is the case, I know that Geometry objects can hold a cache about the pooling operations done to them, so maybe this cache is blowing up the data size. But Pointcept has a data dictionary where just about anything goes inside, and it all gets sent to the GPU memory without any discrimination, so I was hoping that WarpConvNet would be the VRAM-light one.

Functionality wise, they seem to behave similarly just in terms of the loss value at the end of the first epoch repeating the same data sample 100 times. Just a casual observation that I did not verify with a full training experiment on a proper diverse dataset.

**Why still WarpConvNet?**

Even at this stage, I'm still much more interested in using WarpConvNet as opposed to Pointcept. Proper Point and Voxel classes vs statically non-inspectable anything goes inside dictionaries aside, the main weakness of Pointcept is that is is using `spconv`.

For inference on point clouds of large absolute dimensions, even if they can fit inside the GPU no problem, spconv will break. The error message promises a fix in versions 2.2/2.3, but even on these versions the error persists for over a year now https://github.com/traveller59/spconv/issues/706 
The code is so arcane with the author's custom "c++ metaprogramming through python classes" package, that I find it unlikely that any non-maintainer will propose a fix soon either.

With the WarpConvNet version, I can do an inference forward pass (model in eval mode, `torch.no_grad()` context) on a **410m x 410m urban point cloud** with 0.1m voxel size (13M filled voxels!!) **without any tiling**. Of course peak VRAM usage is just as amazing at 86 GB. The largest untiled point cloud I can manage to feed through SparseConv is 330m x 330m, which is still nothing to scoff at (64% of the full point cloud area). Practicality of 0.1m voxel size for urban scale point clouds aside, it is frustrating to be limited not by the hardware, but by the bugs in dependencies.

<img width="389" height="447" alt="Image" src="https://github.com/user-attachments/assets/70ee4745-bca4-4ae5-bcb6-86ca3a5eda49" />

9_Zug_16_34308_-23002 from  [Swiss3DCities](https://zenodo.org/records/4390295)

**Comment by chrischoy** [comment]
Thanks for the analysis. The code is not yet optimized for Hopper and Blackwell since I mostly use Ampere, but will support the newer architectures in the coming months.
